subsys/bluetooth/audio/pbp.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/types.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <stdlib.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/audio/audio.h>
+#include <zephyr/bluetooth/audio/pbp.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/check.h>
+
+LOG_MODULE_REGISTER(bt_pbp, CONFIG_BT_PBP_LOG_LEVEL);
+
+/* PBA Service UUID + Public Broadcast Announcement features + Public_Info metadata of length 1 */
+#define MIN_PBA_SIZE			5
+
+static bool parse_cb(struct bt_data *data, void *user_data)
+{
+	bool *found = (bool *)(user_data);
+
+	if (data->type == BT_AUDIO_METADATA_TYPE_PROGRAM_INFO) {
+		*found = true;
+
+		return false;
+	}
+
+	return true;
+}
+
+int bt_pbp_get_public_broadcast_announcement(const uint8_t meta[], size_t meta_len,
+					     enum bt_pbp_announcement_feature features,
+					     struct net_buf_simple *pba_data_buf)
+{
+	int err;
+	bool program_info_found = false;
+
+	CHECKIF(pba_data_buf == NULL) {
+		LOG_DBG("No buffer provided for advertising data!\n");
+
+		return -EINVAL;
+	}
+
+	CHECKIF((meta == NULL && meta_len != 0) || (meta != NULL && meta_len == 0)) {
+		LOG_DBG("Invalid metadata combination: %p %zu", meta, meta_len);
+
+		return -EINVAL;
+	}
+
+	CHECKIF(pba_data_buf->size < meta_len + 3) {
+		LOG_DBG("Buffer size needs to be at least %d!\n", meta_len + 3);
+
+		return -EINVAL;
+	}
+
+	/* Check that Program Info metadata is included */
+	err = bt_audio_data_parse(meta, meta_len, parse_cb, &program_info_found);
+	CHECKIF(err != 0 && err != -ECANCELED) {
+		LOG_DBG("Could not parse the meta data: %d\n", err);
+		return err;
+	}
```

Comment:
The program info is not strictly necessary as per the PBP spec: 

> A PBS should include the Program_Info length-type-value (LTV) structure metadata

It is "only" a `should`, so it is allowed not to provide it.